### PR TITLE
add-support-for-extra-nic

### PR DESCRIPTION
**Pull request by calvix**
towards https://github.com/giantswarm/giantswarm/issues/7057 and https://github.com/giantswarm/telekom-panamax/issues/43

Sorry for doing more things at once, but I really wanted to clean yet another bunch of useless ancient code from the base. I will highlight the real changes so you can ignore rest.

Changes:
* cleaning of old code that was not used for anything useful
* refactorization of the network config that will include some changes in mayu config file
* adding new options to specify multiple secondary NICs with specific networks and with custom routes

## Important code changes:
adding an entry to save Additional Ips to each host:
https://github.com/giantswarm/mayu/pull/136/files#diff-fb55a563565044be4f4f8f0796c2013aR24

config  and config file refactorization to reuse definitions and remove ancient mysterious naming
https://github.com/giantswarm/mayu/pull/136/files#diff-65381dde95c15e76b01c0767bd11b16eL56-R91

extra nics new template - only single nic support for now
https://github.com/giantswarm/mayu/pull/136/files#diff-115c5b32625e5701a2bab55037a2cfeeR1-R24

generate extraNics ipAddreses when new host/machine is configured
https://github.com/giantswarm/mayu/pull/136/files#diff-26f0bce6f26313b87bfa23cf1f23ce49R70-R74

function for generating a new IP for the extraNic
https://github.com/giantswarm/mayu/pull/136/files#diff-26f0bce6f26313b87bfa23cf1f23ce49R346-R352

a function that ensures all hosts are properly generated IPs for the extraNics
https://github.com/giantswarm/mayu/pull/136/files#diff-26f0bce6f26313b87bfa23cf1f23ce49R372-R405

### CODE REVIEW

File: hostmgr/cluster.go [hunk 1]
```diff
@@ -25,23 +25,14 @@ import (
...
 	cachedModTime time.Time
 	mu            *sync.Mutex
 
-	// indexes
-	hostByInternalAddr map[string]*cachedHost
-	hostByMacAddr      map[string]*cachedHost
-
-	// z
-	predefinedVals map[string]map[string]string
-
```
Comment (calvix) [inline hostmgr/cluster.go]: not really used for anything usefull

File: hostmgr/cluster.go [hunk 2]
```diff
@@ -67,37 +58,25 @@ func OpenCluster(baseDir string, logger micrologger.Logger) (*Cluster, error) {
...
 		}
 	}
 
-	if gitStore && !isGitRepo(baseDir) {
-		err := gitInit(baseDir)
-		if err != nil {
-			return nil, microerror.Mask(err)
-		}
-	}
-
```
Comment (calvix) [inline hostmgr/cluster.go]: git or dropbox store was ancient thing and really useful anymore

File: hostmgr/cluster.go [hunk 3]
```diff
@@ -110,56 +89,28 @@ func NewCluster(baseDir string, gitStore bool, logger micrologger.Logger) (*Clus
 // CreateNewHost creates a new host with the given serial.
 func (c *Cluster) CreateNewHost(serial string) (*Host, error) {
 	serial = strings.ToLower(serial)
-	hostDir := path.Join(c.baseDir, strings.ToLower(serial))
+	hostDir := path.Join(c.baseDir, serial)
```
Comment (calvix) [inline hostmgr/cluster.go]: strings.ToLower was used multiple times on the same string that was already lowered

File: hostmgr/host.go [hunk 4]
```diff
@@ -7,28 +7,27 @@ import (
...
-	Profile          string    `json:",omitempty"`
-	EtcdClusterToken string    `json:",omitempty"`
+	Id               int               `json:",omitempty"`
+	ProviderId       string            `json:",omitempty"`
+	Enabled          bool              `json:",omitempty"`
+	Name             string            `json:",omitempty"`
+	Serial           string            `json:",omitempty"`
+	MacAddresses     []string          `json:",omitempty"`
+	InternalAddr     net.IP            `json:",omitempty"`
+	AdditionalAddrs  map[string]net.IP `json:",omitempty"`
```
Comment (calvix) [inline hostmgr/host.go]: adding new entry here `AdditionalAddrs  map[string]net.IP` will be a map with `interface-name` : `assigned IP `

File: pxemgr/config.go [hunk 5]
```diff
@@ -53,29 +53,47 @@ type Profile struct {
...
+
+	PrimaryNIC NetworkInterface   `yaml:"primary_nic"`
+	ExtraNICs  []NetworkInterface `yaml:"extra_nics"`
+
+	// if set true use UEFI boot, otherwise use legacy BIOS
+	UEFI bool
+
+	// DNS and NTP list for installed machines
+	DNS []string
+	NTP []string
```
Comment (calvix) [inline pxemgr/config.go]: there is a refactoring of the Network config structure to remove some ancient and  misleading naming and unify it across all resources

File: pxemgr/config.go [hunk 6]
```diff
@@ -53,29 +53,47 @@ type Profile struct {
...
 	BondInterfaceMatch string `yaml:"bond_interface_match"`
+}
+
+type NetworkInterface struct {
+	Routes        []NetworkRoute `yaml:"routes"`
+	InterfaceName string         `yaml:"interface_name"`
+	IPRange       NetworkRange   `yaml:"ip_range"`
+	SubnetSize    string         `yaml:"subnet_size"`
+	SubnetGateway string         `yaml:"subnet_gateway"`
+}
```
Comment (calvix) [inline pxemgr/config.go]: the network interface will now be used for defining: primary NIC, pxe NIC, and the extra NICs, all of them allow specify extra custom routes for each interface

File: pxemgr/key.go [hunk 7]
```diff
@@ -1,43 +1,19 @@
...
 )
 
 const (
-	vmlinuzFile      = "coreos_production_pxe.vmlinuz"
-	initrdFile       = "coreos_production_pxe_image.cpio.gz"
-	installImageFile = "coreos_production_image.bin.bz2"
-	qemuImageFile    = "coreos_production_qemu_usr_image.squashfs"
-	qemuKernelFile   = "coreos_production_qemu.vmlinuz"
+	vmlinuzFile = "coreos_production_pxe.vmlinuz"
+	initrdFile  = "coreos_production_pxe_image.cpio.gz"
```
Comment (calvix) [inline pxemgr/key.go]: removing, we no more pull qemu images from the mayu file server, nor install  coreos into machines

pxe images are enough

File: pxemgr/pxemanager.go [hunk 8]
```diff
@@ -195,6 +195,9 @@ func PXEManager(c PXEManagerConfiguration, cluster *hostmgr.Cluster) (*pxeManage
 		mgr.etcdDiscoveryUrl = mgr.config.TemplatesEnv["mayu_https_endpoint"].(string) + "/etcd"
 	}
 
+	// we need to do this on boot time to ensure all newly added Network.ExtraNICs have properly assigned IP to all hosts
+	mgr.checkAdditionalNICAddresses()
+
```
Comment (calvix) [inline pxemgr/pxemanager.go]: this is an important part of code which on boot checks if proper IPs are assigned to all Extra NICs as there can be scenario when customer want to add new NICs to already installed cluster

implementation is lower in the diff

File: pxemgr/pxemanager.go [hunk 9]
```diff
@@ -240,12 +243,9 @@ func (mgr *pxeManagerT) startAPIserver() error {
 	mgr.apiRouter = mux.NewRouter()
 	//  api endpoint for setting metadata of machine
 	mgr.apiRouter.Methods("PUT").PathPrefix("/admin/host/{serial}/boot_complete").HandlerFunc(withSerialParam(mgr.bootComplete))
-	mgr.apiRouter.Methods("PUT").PathPrefix("/admin/host/{serial}/mark_fresh").HandlerFunc(withSerialParam(mgr.markFresh))
 	mgr.apiRouter.Methods("PUT").PathPrefix("/admin/host/{serial}/set_provider_id").HandlerFunc(withSerialParam(mgr.setProviderId))
 	mgr.apiRouter.Methods("PUT").PathPrefix("/admin/host/{serial}/set_ipmi_addr").HandlerFunc(withSerialParam(mgr.setIPMIAddr))
-	mgr.apiRouter.Methods("PUT").PathPrefix("/admin/host/{serial}/set_state").HandlerFunc(withSerialParam(mgr.setState))
 	mgr.apiRouter.Methods("PUT").PathPrefix("/admin/host/{serial}/set_etcd_cluster_token").HandlerFunc(withSerialParam(mgr.setEtcdClusterToken))
-	mgr.apiRouter.Methods("PUT").PathPrefix("/admin/host/{serial}/override").HandlerFunc(withSerialParam(mgr.override))
```
Comment (calvix) [inline pxemgr/pxemanager.go]: endpoints that were not really useful to anything and were artifacts from ancient times

File: pxemgr/ipxe_handlers.go [hunk 10]
```diff
@@ -65,56 +65,33 @@ func (mgr *pxeManagerT) maybeCreateHost(serial string) (*hostmgr.Host, error) {
...
-				return nil, microerror.Mask(err)
-			}
 		}
-
+		// generate addresses for the extra NICs
+		extraNicCount := len(mgr.config.Network.ExtraNICs)
+		host.AdditionalAddrs = make(map[string]net.IP)
+		for i := 0; i < extraNicCount; i++ {
+			host.AdditionalAddrs[mgr.config.Network.ExtraNICs[i].InterfaceName] = mgr.getNextAdditionalIP(i, mgr.config.Network.ExtraNICs[i].InterfaceName)
+		}
```
Comment (calvix) [inline pxemgr/ipxe_handlers.go]: assign additional IP to ExtraNICs when new machine is added to the configuration

File: pxemgr/ipxe_handlers.go [hunk 11]
```diff
@@ -65,56 +65,33 @@ func (mgr *pxeManagerT) maybeCreateHost(serial string) (*hostmgr.Host, error) {
...
 				host.Profile = defaultProfileName
 			}
-			host.CoreOSVersion = mgr.profileCoreOSVersion(host.Profile)
-			host.EtcdClusterToken = mgr.profileEtcdClusterToken(host.Profile)
-
-			err = host.Commit("updated host profile and metadata")
-			if err != nil {
-				return nil, microerror.Mask(err)
-			}
+			host.CoreOSVersion = mgr.config.DefaultCoreOSVersion
```
Comment (calvix) [inline pxemgr/ipxe_handlers.go]: we don't really support several different CoreOS version across installation so we use the default one everywhere

File: pxemgr/ipxe_handlers.go [hunk 12]
```diff
@@ -217,21 +179,6 @@ func setContentLength(w http.ResponseWriter, f *os.File) error {
...
-		return
-	}
-
-	host.State = hostmgr.Configured
-	host.Commit("host flagged as fresh")
-	mgr.cluster.Update()
-
-	w.WriteHeader(202)
-}
-
```
Comment (calvix) [inline pxemgr/ipxe_handlers.go]: useless

File: pxemgr/ipxe_handlers.go [hunk 13]
```diff
@@ -349,129 +289,110 @@ func (mgr *pxeManagerT) setState(serial string, w http.ResponseWriter, r *http.R
...
+		currentIP = incIP(currentIP)
+	}
+}
 
-		if host.CoreOSVersion == "" {
-			return mgr.config.DefaultCoreOSVersion
-		} else {
-			return host.CoreOSVersion
+// check af all hosts have properly assigned IP addresses to all Network.ExtraNICs entries
+func (mgr *pxeManagerT) checkAdditionalNICAddresses() {
```
Comment (calvix) [inline pxemgr/ipxe_handlers.go]: this is an important part of code which on boot checks if proper IPs are assigned to all Extra NICs as there can be a scenario when customer want to add new NICs to already installed cluster or the range could change

**Comment by calvix** [comment]
Code looks good, starting with proper testing

**Comment by calvix** [comment]
example of mayu config with new format with example of use extraNICs and routes

```yaml
default_coreos_version: 2303.3.0
network:
  bind_addr: 10.0.4.250
  pxe:
    enabled: true
    pxe_interface:
      interface_name: eth0
      ip_range:
        start: 10.0.4.10
        end: 10.0.4.40
      subnet_gateway: 10.0.4.254
      subnet_size: 24
  primary_nic:
    interface_name: eth0
    ip_range:
      start: 10.0.4.101
      end: 10.0.4.200
    subnet_size: 24
    subnet_gateway: 10.0.4.254
    network_model:
      type: singlenic
    routes:
    - destination_cidr: 11.12.13.0/24
      route_hop: 10.0.4.200
    - destination_cidr: 87.12.13.0/28
      route_hop: 10.0.4.199
  extra_nics:
  - interface_name: eth1
    ip_range:
      start: 10.0.5.210
      end: 10.0.5.245
    subnet_size: 24
    network_model:
      type: singlenic
    routes:
      - destination_cidr: 1.1.4.0/28
        route_hop: 10.0.5.200
      - destination_cidr: 8.123.10.0/29
        route_hop: 10.0.5.121
  - interface_name: eth2
    ip_range:
      start: 10.0.6.210
      end: 10.0.6.245
    subnet_size: 24
    network_model:
      type: singlenic
  dns:
  - 1.1.1.1
  - 8.8.4.4
  ntp:
  - 146.0.32.144
  - 129.70.132.33
  - 62.116.162.126
  uefi: false

```

File: pxemgr/iputil.go [hunk 14]
```diff
@@ -16,18 +19,10 @@ func incIP(ip net.IP) net.IP {
...
-	}
-	if ip[2] < upperBound[2] {
-		return true
-	}
-	if ip[1] < upperBound[1] {
-		return true
-	}
-	if ip[0] <= upperBound[0] {
-		return true
-	}
```
Comment (calvix) [inline pxemgr/iputil.go]: this actually never worked :p no idea who wrote this code but the test file was working becasue there was only the cases that were suppsoe to be `true` and the function always returned `true` no matter what input

**Comment by mazzy89** [comment]
So hard to review :-(

**Comment by calvix** [comment]
@mazzy89  I know, that's why I added the  summary where all the new code is added in the issue description

If its still not enought  for everyone I will try to make a separate PR than

**Review by mazzy89 — APPROVED** [review]
The model makes sense to me.

File: pxemgr/config.go (hunk 5 above)
Comment (corest) [reply pxemgr/config.go]: I see there is single DNS field for all interfaces. Would be neat to have it per interface if that's specified

Comment (calvix) [reply pxemgr/config.go]: I don't think `systemd-networkd` support this or how would that work? How is OS deciding which DNS should it use?

Comment (calvix) [reply pxemgr/config.go]: Looks like systemd-resolved supports it

Comment (calvix) [reply pxemgr/config.go]: adding

Comment (calvix) [reply pxemgr/config.go]: done, moved DNS config under network interface

PXE is using  primary NICs DNS so we dont have to define it multiple times

**Pull request merged**